extendr-api/src/thread_safety.rs

```diff
@@ -1,69 +1,47 @@
 //! Provide limited protection for multithreaded access to the R API.
 
 use crate::*;
-use std::sync::atomic::{AtomicU32, Ordering};
+use std::cell::Cell;
+use std::sync::Mutex;
 
-static OWNER_THREAD: AtomicU32 = AtomicU32::new(0);
-static NEXT_THREAD_ID: AtomicU32 = AtomicU32::new(1);
+/// A global lock, that should represent the global lock on the R-API.
+/// It is not tied to an actual instance of R.
+static R_API_LOCK: Mutex<()> = Mutex::new(());
 
 thread_local! {
-    static THREAD_ID: u32 = NEXT_THREAD_ID.fetch_add(1, Ordering::SeqCst);
+    static THREAD_HAS_LOCK: Cell<bool> = Cell::new(false);
 }
 
-// Get an integer 1.. for each thread that calls this.
-pub fn this_thread_id() -> u32 {
-    THREAD_ID.with(|&v| v)
-}
-
-/// Run a function single threaded.
-/// Note: This will fail badly if the called function panics or calls RF_error.
+/// Run `f` while ensuring that `f` runs in a single-threaded manner.
 ///
-/// ```
-/// use extendr_api::prelude::*;
-/// use std::sync::atomic::{AtomicU32, Ordering};
-/// static GLOBAL_THREAD_COUNT: AtomicU32 = AtomicU32::new(0);
+/// This is intended for single-threaded access of the R's C-API.
+/// It is possible to have nested calls of `single_threaded` without deadlocking.
 ///
-/// let threads : Vec<_> = (0..100).map(|_| {
-///    std::thread::spawn(move|| {
-///       single_threaded(|| {
-///         // check that we are single threaded.
-///         let old_thread_count = GLOBAL_THREAD_COUNT.fetch_add(1, Ordering::AcqRel);
-///         assert_eq!(old_thread_count, 0);
-///         std::thread::sleep(std::time::Duration::from_millis(1));
-///         GLOBAL_THREAD_COUNT.fetch_sub(1, Ordering::AcqRel);
-///         // recursive calls are ok.
-///         assert_eq!(single_threaded(|| {
-///           1
-///         }), 1);    
-///       })
-///    })
-/// }).collect();
-/// ```
+/// Note: This will fail badly if the called function `f` panics or calls `Rf_error`.
 pub fn single_threaded<F, R>(f: F) -> R
 where
     F: FnOnce() -> R,
 {
-    let id = this_thread_id();
-    let old_id = OWNER_THREAD.load(Ordering::Acquire);
-
-    if old_id != id {
-        // wait for OWNER_THREAD to become 0 and put us as the owner.
-        while OWNER_THREAD
-            .compare_exchange(0, id, Ordering::Acquire, Ordering::Relaxed)
-            .is_err()
-        {
-            std::thread::sleep(std::time::Duration::from_millis(1));
-        }
-    }
+    let has_lock = THREAD_HAS_LOCK.with(|x| x.get());
+
+    // acquire R-API lock
+    let _guard = if !has_lock {
+        Some(R_API_LOCK.lock().unwrap())
+    } else {
+        None
+    };
+
+    // this thread now has the lock
+    THREAD_HAS_LOCK.with(|x| x.set(true));
 
-    let res = f();
+    let result = f();
```

Comment:
Last question -- do we want to handle mutex poisoning somehow?

Reply:
Ohhh...... interesting.. Future PR?

Reply:
I am new to Rust mutex, but from what I've read, if it gets poisoned, currently it will basically nuke the whole process because no other thread will be able to enter it, ever.